### PR TITLE
feat: tracking basic events

### PR DESCRIPTION
**Pull request by jjmachan**
This PR implement event tracking for yatai.

To envvars to control tracking
```
export __YATAI_TRACKING_LOGLEVEL=[debug|info]
export YATAI_DONOT_TRACK=true
```

by default, the lifecycle event runs every 6 hours. If `__YATAI_TRACKING_LOGLEVEL='debug` then yatai tracking is in debug mode and the lifecycle event runs every minute

### CODE REVIEW

File: common/tracking/tracker.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+const (
+	JITSU_SERVER              = "http://localhost:8000/api/v1/s2s/event"
+	JITSUE_KEY                = "s2s.l5z3kpmcuevukrepeg70ht.j2imnduhnssuh4yb7gt9"
+	YATAI_TRACKING_DEBUG_MODE = "__YATAI_TRACKING_DEBUG_MODE"
+	YATAI_DONOT_TRACK         = "YATAI_DONOT_TRACK"
+)
+
+func is_debug_mode() bool {
+	out := os.Getenv(YATAI_TRACKING_DEBUG_MODE)
+	if strings.ToLower(out) == "true" {
```
Comment (yetone) [inline common/tracking/tracker.go:28]:
```suggestion
	return strings.ToLower(out) == "true"
```

File: common/tracking/tracker.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+		return true
+	} else {
+		return false
+	}
+}
+
+func donot_track() bool {
+	out := os.Getenv(YATAI_DONOT_TRACK)
+	if strings.ToLower(out) == "true" {
```
Comment (yetone) [inline common/tracking/tracker.go:37]:
```suggestion
	return strings.ToLower(out) == "true"
```

File: common/tracking/schemas.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	Timestamp       time.Time
+	YataiVersion    string
+}
+
+type TriggerEvent struct {
+	UserUID string
+}
+
+type LifecycleEvent struct {
+	LifecycleEventType string
```
Comment (yetone) [inline common/tracking/schemas.go:20]: It is recommended that each schema structure field should have `json` tags

Comment (jjmachan) [reply common/tracking/schemas.go]: thank you for the tip :) added it

File: api-server/cmd/serve.go [hunk 4]
```diff
@@ -126,6 +127,7 @@ func (opt *ServeOption) Run(ctx context.Context, args []string) error {
 	}
 
 	addCron(ctx)
+	tracking.AddLifeCycleTrackingCron(ctx)
```
Comment (yetone) [inline api-server/cmd/serve.go:130]: can you put this calling into addCron?

File: api-server/services/tracking/lifecycle_events.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+	"github.com/bentoml/yatai/api-server/version"
+	"github.com/tianweidut/cron"
+)
+
+func AddLifeCycleTrackingCron(ctx context.Context) {
+	TrackLifeCycle(ctx, YataiLifeCycleStartup)
+	ctx = context.WithValue(ctx, "yataiUpTimestamp", time.Now())
+
+	c := cron.New()
+	err := c.AddFunc("@every 1m", func() {
```
Comment (yetone) [inline api-server/services/tracking/lifecycle_events.go:18]: Can you make this period configurable?

File: api-server/services/tracking/lifecycle_events.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+		TrackLifeCycle(ctx, YataiLifeCycleUpdate)
```
Comment (yetone) [inline api-server/services/tracking/lifecycle_events.go:19]:
```go
ctx_, cancel = context.WithTimeout(ctx, time.Minute)
defer cancel()
TrackLifeCycle(ctx_, ...)
```

File: api-server/services/tracking/lifecycle_events.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+import (
+	"context"
+	"time"
+
+	"github.com/bentoml/yatai-schemas/modelschemas"
+	"github.com/bentoml/yatai/api-server/services"
+	"github.com/bentoml/yatai/api-server/version"
+	"github.com/tianweidut/cron"
+)
```
Comment (yetone) [inline api-server/services/tracking/lifecycle_events.go:11]:
```go
var (
    yataiUpTimestamp time.Time
    yataiUpTimestampMu sync.Mutex
)

func resetYataiUpTimestamp() {
    yataiUpTimestamp.Lock()
    defer yataiUpTimestamp.Unlock()
    yataiUptimestamp = time.Now()
}

func init() {
    resetYataiUpTimestamp()
}

File: api-server/services/tracking/tracker.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+		logLevel = log.DebugLevel
+	default:
+		logLevel = log.FatalLevel
+	}
+	logger := log.New()
+	logger.SetLevel(logLevel)
+	return logger
+}
+
+func donot_track() bool {
```
Comment (yetone) [inline api-server/services/tracking/tracker.go:41]: need to be camelCase

File: api-server/services/tracking/tracker.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+		trackingLogger.Error(err, "Failed to marshal data")
+		return
+	}
+
+	var prettyJSON bytes.Buffer
+	_ = json.Indent(&prettyJSON, jsonData, "", " ")
+	trackingLogger.Info("Tracking Payload: ", prettyJSON.String())
+
+	if !donot_track() {
+		client := http.Client{Timeout: TRACKER_TIMEOUT}
```
Comment (yetone) [inline api-server/services/tracking/tracker.go:61]: use `yatai-common` to reuse http client

**Comment by yetone** [comment]
please fix the golint

File: api-server/controllers/controllersv1/deployment.go [hunk 10]
```diff
@@ -421,7 +426,9 @@ func (c *deploymentController) Delete(ctx *gin.Context, schema *GetDeploymentSch
 	if err != nil {
 		return nil, err
 	}
-	return transformersv1.ToDeploymentSchema(ctx, deployment)
+	deploymentSchema, err := transformersv1.ToDeploymentSchema(ctx, deployment)
+	go tracking.TrackDeploymentEvent(ctx, deploymentSchema, tracking.YataiDeploymentDelete)
```
Comment (aarnphm) [inline api-server/controllers/controllersv1/deployment.go:430]: qq: why does this has to be a goroutine?

Comment (jjmachan) [reply api-server/controllers/controllersv1/deployment.go:430]: didn't want to make the tracking requests blocking which is why we are spinning off a goroutine to every tracking request

File: helm/yatai/templates/deployment.yaml [hunk 11]
```diff
@@ -67,6 +67,8 @@ spec:
                   name: {{ .Values.s3.secretKeyExistingSecretName | quote }}
                   key: {{ .Values.s3.secretKeyExistingSecretKey | quote }}
             {{- end }}
+            - name: YATAI_DONOT_TRACK
+              value: {{ .Values.doNotTrack | quote }}
```
Comment (jjmachan) [inline helm/yatai/templates/deployment.yaml:71]: setting do not track directly in deployment

**Review by yetone — APPROVED** [review]

**Pull request merged**